### PR TITLE
Add owners command

### PR DESCRIPTION
**Pull request by doug-wade**
Addresses #38

### CODE REVIEW

File: bin/lerna.js [hunk 1]
```diff
@@ -20,6 +20,7 @@ var cli = meow([
   "  run        Run npm script in each package",
   "  exec       Run a command in each package",
   "  ls         List all public packages",
+  "  owners     Add and remove package owners",
```
Comment (gigabo) [inline bin/lerna.js]: Might be nice to make it `owner` to be consistent with the `npm owner` command.

File: src/ConfigUtilities.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  static read(rootPath) {
+    config = config || require(path.join(rootPath, "lerna.json"));
+    return config;
+  }
+
+  @logger.logifySync()
+  static write(rootPath, conf) {
+    config = conf;
+    FileSystemUtilities.writeFileSync(path.join(rootPath, "lerna.json"), JSON.stringify(config));
+  }
```
Comment (gigabo) [inline src/ConfigUtilities.js]: Nice that these are encapsulated so they can be stubbed for tests.

File: src/commands/OwnersCommand.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  }
+
+  runCommandInPackage(pkg) {
+    const pkgName = pkg._package.name;
+    if (this.command === "add") {
+      NpmUtilities.addOwner(this.ownerName, pkgName);
+      this.logger.info(`Added owner ${this.ownerName} to package ${pkgName}`);
+    } else if (this.command === "rm") {
+      NpmUtilities.removeOwner(this.ownerName, pkgName);
+      this.logger.info(`Removed owner ${this.ownerName} to package ${pkgName}`);
```
Comment (gigabo) [inline src/commands/OwnersCommand.js]: ... _from_ package?

File: src/commands/OwnersCommand.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    }
+    config.owners = [...ownersSet];
+    ConfigUtilities.write(this.repository.rootPath, config);
+    async.parallelLimit(this.filteredPackages.map((pkg) => (cb) => {
+      this.runCommandInPackage(pkg, cb);
+    }), this.concurrency, callback);
+  }
+
+  runCommandInPackage(pkg) {
+    const pkgName = pkg._package.name;
```
Comment (gigabo) [inline src/commands/OwnersCommand.js]: There's actually [a `name` getter](https://github.com/lerna/lerna/blob/master/src/Package.js#L13-L15), so you should be able to just do `pkg.name` here.

**Comment by doug-wade** [comment]
Thanks for the quick response @gigabo.  I updated the docs and added tests, and did a little general cleanup and commenting.

**Comment by doug-wade** [comment]
rebased onto the latest master, and got the tests running -- wanna take a look @gigabo?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -19,6 +19,14 @@ Or the linter:
 $ npm run lint
 ```
 
+If you want to run a single unit test, you can globally install the test runner
+and point it at the file you're interested in
+
+```sh
+$ npm i -g mocha
+$ mocha -t 5000 path/to/your/test.js
```
Comment (gigabo) [inline CONTRIBUTING.md]: Oh, I've just been doing `$(npm bin)/mocha path/to/test.js`.  Guess that's the same.

Nice addition to the docs!

File: src/index.js [hunk 6]
```diff
@@ -2,24 +2,26 @@ import BootstrapCommand from "./commands/BootstrapCommand";
...
   import: ImportCommand,
+  owner: OwnerCommand,
   clean: CleanCommand,
   diff: DiffCommand,
   init: InitCommand,
-  run: RunCommand,
   exec: ExecCommand,
-  ls: LsCommand
+  run: RunCommand,
+  ls: LsCommand,
```
Comment (gigabo) [inline src/index.js]: Heads up: This is going to conflict with the [automatic hoisting PR](https://github.com/lerna/lerna/pull/507/files#diff-1fdf421c05c1140f6d71444ea2b27638).

Might help the merge to trim it down to just the add of `owner` for now and handle the reorder separately.

File: src/commands/OwnerCommand.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    // won't be in the configured registry.
+    async.parallelLimit(this.filteredPackages.map((pkg) => (cb) => {
+      this.runCommandInPackage(pkg, cb);
+    }), this.concurrency, callback);
+  }
+
+  runCommandInPackage(pkg, callback) {
+    switch (this.command) {
+    case ADD_COMMAND_NAME:
+      NpmUtilities.addOwner(this.ownerName, pkg.name);
```
Comment (gigabo) [inline src/commands/OwnerCommand.js]: No trouble with this if called with a package/owner pair that's already set up?  Same with remove, below?

File: CONTRIBUTING.md (hunk 5 above)
Comment (xaka) [inline CONTRIBUTING.md]: It's kind of dangerous to install mocha in such way as you might end up with version different from what `lerna` uses? It also leads to pollution of global namespace. As `mocha` is already part of `devDependencies` and aliased via `test` script, you can do the following:

```bash
npm run test -- <path> # or any other cmd arg mocha understands
```

File: src/commands/OwnerCommand.js [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+    switch (this.command) {
+    case ADD_COMMAND_NAME:
+      ownersSet.add(this.ownerName);
+      break;
+    case RM_COMMAND_NAME:
+      if (ownersSet.has(this.ownerName)) {
+        ownersSet.delete(this.ownerName);
+      } else {
+        return new callback(new Error(this.ownerName + OWNER_NOT_FOUND_ERROR_FRAGMENT));
```
Comment (gigabo) [inline src/commands/OwnerCommand.js]: Should this warn and proceed?  Could be trying to remove an owner that was added before owners were stored in lerna config.

File: CONTRIBUTING.md (hunk 5 above)
Comment (gigabo) [inline CONTRIBUTING.md]: Ooh, nice tip @xaka!

File: src/commands/OwnerCommand.js (hunk 7 above)
Comment (doug-wade) [inline src/commands/OwnerCommand.js]: not that I can tell -- when I add myself as an owner repeatedly or remove an owner from a package I own repeatedly, npm doesn't throw or log an error or anything.

File: src/commands/OwnerCommand.js (hunk 8 above)
Comment (doug-wade) [inline src/commands/OwnerCommand.js]: oooh good call 👍

**Comment by gigabo** [comment]
Are the failing tests here just waiting for a rebase onto #557?

**Comment by doug-wade** [comment]
@gigabo yes

**Pull request closed without merging**